File: frontend/network_layout.py
"""
A datastructure that models the visual layout of the NN model.

Positions are within the unit square

Initially I thought of using networkx (like in some tutorials), but it doesn't
properly capture the layered shape of neuron network, nor does it know how to layout
linearly top to bottom
"""
from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class Point:
    x: float
    y: float


@dataclass(frozen=True)
class LayerSpec:
    name: str
    neuron_count: int
    neurons_per_row: Optional[int] = None


@dataclass()
class Layer:
    name: str
    neuron_count: int
    neurons_per_row: Optional[int]
    rank: int
    total_layer_count: int
    neuron_activations = Optional[list[float]]

    def __post_init__(self):
        if self.neurons_per_row is None:
            self.neurons_per_row = self.neuron_count

        self.neuron_positions = []
        y = 1 - self.rank / self.total_layer_count
        index_in_row = 0
        for _ in range(self.neuron_count):
            self.neuron_positions.append(
                Point(index_in_row / (self.neurons_per_row - 1), y)
            )
            index_in_row += 1
            if index_in_row == self.neurons_per_row:
                index_in_row = 0
                y += 0.07  # XXX
        self.neuron_activations = None

    @property
    def get_neuron_activations(self) -> list[float]:
        if self.neuron_activations is None:
            self.neuron_activations = [0] * self.neuron_count
        return self.neuron_activations


class Network:
    def __init__(self, *layer_specs: LayerSpec):
        self.weights: list[list[float]] = []
        self.layers: list[Layer] = []
        for i, spec in enumerate(layer_specs):
            layer = Layer(
                name=spec.name,
                neuron_count=spec.neuron_count,
                neurons_per_row=spec.neurons_per_row,
                rank=i,
                total_layer_count=len(layer_specs),
            )
            self.layers.append(layer)

    def set_activations(self, activations: list[list[float]]) -> None:
        for layer, layer_activations in zip(self.layers, activations):
            layer.neuron_activations = layer_activations

File: frontend/src/lib/network-layout.ts
// Models the visual layout of a fully connected feed forward neural network
// Positions are within the unit square.

import type { LayerVariable } from '@tensorflow/tfjs';
import { zip } from './pythonish';

export class Point {
	x: number;
	y: number;

	constructor(x: number, y: number) {
		this.x = x;
		this.y = y;
	}
}

export class Neuron {
	x: number;
	y: number;
	activation: number | null;

	constructor(x: number, y: number, activation: number | null) {
		this.x = x;
		this.y = y;
		this.activation = activation;
	}
}

export class Link {
	a: Neuron;
	b: Neuron;
	weight: number | null;

	constructor(a: Neuron, b: Neuron, weight: number | null) {
		this.a = a;
		this.b = b;
		this.weight = weight;
	}
}

export class LayerSpec {
	name: string;
	neuron_count: number;
	marker_size: number;
	neurons_per_row: number | null;

	constructor(
		name: string,
		neuron_count: number,
		marker_size: number,
		neurons_per_row: number | null = null
	) {
		this.name = name;
		this.neuron_count = neuron_count;
		this.marker_size = marker_size;
		this.neurons_per_row = neurons_per_row;
	}
}

export class Layer {
	name: string;
	marker_size: number;
	neurons: Neuron[];

	constructor(rank: number, total_layer_count: number, spec: LayerSpec) {
		this.name = spec.name;
		this.marker_size = spec.marker_size;
		const neuron_count = spec.neuron_count;
		const neurons_per_row = spec.neurons_per_row || spec.neuron_count;

		const rows_in_layer = Math.floor(neuron_count / neurons_per_row);
		let y = 1 - rank / total_layer_count + rows_in_layer * 0.02;

		const neuron_positions = [];
		let index_in_row = 0;
		for (let i = 0; i < neuron_count; i++) {
			neuron_positions.push(new Point(index_in_row / (neurons_per_row - 1), y));
			index_in_row++;
			if (index_in_row == neurons_per_row) {
				index_in_row = 0;
				y -= 0.02;
			}
		}

		this.neurons = neuron_positions.map((pos) => new Neuron(pos.x, pos.y, 0));
	}
}

export type LinkFilter = (link: Link) => boolean;

export function allLinks(_: Link) {
	return true;
}

export class DenseNetwork {
	layers: Layer[];

	constructor(layer_specs: LayerSpec[]) {
		this.layers = [];
		for (const [rank, spec] of layer_specs.entries()) {
			this.layers.push(new Layer(rank, layer_specs.length, spec));
		}
	}

	links(weights: LayerVariable[], linkFilter: LinkFilter) {
		const links: Link[] = [];

		// console.log(zip(network.layers.slice(0, -1), network.layers.slice(1), weights));

		for (const [from_layer, to_layer, weights_between_layers_tensor] of zip(
			this.layers.slice(0, -1),
			this.layers.slice(1),
			weights
		)) {
			const weights_between_layers = weights_between_layers_tensor.val.arraySync();
			// console.log("from layer", from_layer);
			// console.log("to layer", to_layer);
			// console.log("weights between layers", weights_between_layers);

			for (const [from_neuron, outgoing_weights_for_neuron] of zip(
				from_layer.neurons,
				weights_between_layers
			)) {
				// console.log("from neuron", from_neuron);
				// console.log("outgoing weights for neuron", outgoing_weights_for_neuron);

				for (const [to_neuron, weight] of zip(to_layer.neurons, outgoing_weights_for_neuron)) {
					// console.log('from', from_neuron, 'to', to_neuron, 'weight', weight);
					const link = new Link(from_neuron, to_neuron, weight);
					if (linkFilter(link)) {
						links.push(link);
					}
				}
			}
		}

		return links;
	}
}
